### PR TITLE
doc: capture since frontiers in correctness document

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

Precisely capture the property that we require for https://github.com/MaterializeInc/materialize/issues/17643
### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: doc/developer/platform/ux.md [hunk 1]
```diff
@@ -905,6 +905,12 @@ Materialize maintains four correctness guarantees.
    Materialize sends an acknowledgement for (*e.g., the commit of an offset by a
    Kafka source, the acknowledgement of an LSN by a PostgresSQL source, etcetera*),
    will be saved in durable storage.
+5. *Materialize respects the compaction frontier of its sources.* This means
+   that on source creation Materialize queries and puts a read hold on the
+   `initial_since` frontier of the upstream system and sets the initial since
+   *of the collection in Materialize* to be `{reclock(t) for t in
+   initial_since}`. Informally, this prevents users from reading a collection
+   for a source where the contents describe a state that never happened.
```
Comment (benesch) [inline doc/developer/platform/ux.md:913]: Might be worth explicitly discussing that not all sources support read holds, and in that case Materialize just has to make its best attempt.

**Review by benesch — COMMENTED** [review]
This makes sense to me, but I haven't thought hard enough myself about whether something about (2) already implies this. Hopefully @frankmcsherry has!

File: doc/developer/platform/ux.md (hunk 1 above)
Comment (petrosagg) [reply doc/developer/platform/ux.md:913]: Given that having upstream compaction run ahead under your feet causes correctness issues what should sources do when their best attempt fails? In other words, should the kafka source immediately emit a permanent error if it detects upstream compaction went beyond its read cursor?

Comment (benesch) [reply doc/developer/platform/ux.md:913]: Is that even detectable? I was just thinking that the correct property should leave itself an exemption for sources that don't support read holds, and then we can figure out what the best effort behavior is on a source by source basis.

"Materialize respects the compaction frontier of its sources, to the extent permitted by the upstream system."

Comment (petrosagg) [reply doc/developer/platform/ux.md]: > Is that even detectable? 

Depends on the source. A postgres source with a non-zero `max_slot_wal_keep_size` will advance the since frontier past us if we stay down long enough and it is detectable. For Kafka I'm not sure. I'll rephrase it to leave it up to source implementations although it would be nice to have a preferred behavior under the assumption that it is detectable.

**Comment by frankmcsherry** [comment]
Mentioned in DM and maybe we'll copy/paste here, but I'd love to understand the alternate strategy of modifying correctness property 2 to speak about sequences of upstream data state. E.g. we lean in to "all sources are CDC" as a framing, and property 2 is "we present a sequence of upstream collections that did occur, and in that order".

The issue linked points back at the discussion here, but tbh I don't see the explanation for this and at the very least we should record it (and ideally I'd love to understand it too but let's mark that as optional).
